### DISCUSSION TITLE
Post route not found

### DISCUSSION THREAD

**Discussion by rubypirate** [naymspace/backpex#644] [Q&A]
# Missing POST Route for Form Submission in Backpex Admin

I'm encountering a routing issue while trying to create a new post using Backpex admin interface. Here are the details:

## Environment
- Elixir: 1.14
- Phoenix: 1.7.14
- LiveView: 1.0.0-rc.1
- Backpex: 0.7.1

## Router
```elixir
defmodule AppWeb.Router do
  use AppWeb, :router

  import Backpex.Router
  import AppWeb.UserAuth

  pipeline :browser do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_live_flash
    plug :put_root_layout, html: {AppWeb.Layouts, :root}
    plug :protect_from_forgery
    plug :put_secure_browser_headers
    plug :fetch_current_user
    plug Backpex.ThemeSelectorPlug
  end

  pipeline :api do
    plug :accepts, ["json"]
  end

  scope "/", AppWeb do
    pipe_through :browser

    get "/", PageController, :home
  end

  scope "/admin", AppWeb do
    pipe_through :browser
    backpex_routes()

    live_session :default, on_mount: Backpex.InitAssigns do
      live_resources "/posts", Live.PostLive
    end
  end
end

```

## Steps to Reproduce
1. Created a new Phoenix application 
2. Set up Backpex following the installation guide
3. Navigated to `/admin/posts` (displays correctly)
4. Attempted to submit a new post form

## Current Behavior
Receiving error: `no route found for POST /admin/posts/new (AppWeb.Router)`

## Expected Behavior
Form should submit successfully and create a new post.

## Additional Information
- Verified all configuration steps from the installation guide
- Tested with the official Backpex demo repository and encountered the same issue

### Available Routes
```
POST  /admin/backpex_cookies                    Backpex.CookieController :update
GET   /admin/posts                              DemoWeb.PostLive :index
GET   /admin/posts/new                          DemoWeb.PostLive :new
GET   /admin/posts/:backpex_id/edit             DemoWeb.PostLive :edit
GET   /admin/posts/:backpex_id/show             DemoWeb.PostLive :show
```

Any assistance would be appreciated.

**Comment by Flo0807** [comment]
Hey!

I cannot find any obvious errors in your setup. I have also checked the demo application (`main` and `develop`). Everything seems to work fine there.

From your list of available routes, the `GET /admin/posts/new` route is generated properly, too. The problem seems to be that it tries to call the `POST` route instead of `GET`.

I can only guess that it's caused by LiveView. Maybe try to use the latest official supported LiveView version instead of the release candidate. We did not test the compatibility with LiveView 1.0 yet.

You mention that the error occurs when creating a new LiveView application with a basic Backpex setup. Therefore, if the error still occurs, maybe you can share a minimal repo (?) Then I can look into it in detail.

**Reply by rubypirate** [reply]
Thanks @Flo0807 , I used latest backpex version and everything worked just fine.

**Comment by Hkalani** [comment]
I'm also facing this issue, (given i was following an elixir cast video on backpex 0.6.0, I did the migration to 0.8.0. ) Everything works except any POST actions I tried. 
![Screenshot 2024-11-11 123615](https://github.com/user-attachments/assets/90f9ccb1-fde2-4c8f-9b7a-ab276fa27aaa)
![Screenshot 2024-11-11 123556](https://github.com/user-attachments/assets/ba5df250-bfea-4f75-9009-19f60772d794)
![Screenshot 2024-11-11 123549](https://github.com/user-attachments/assets/6d67094a-217e-4d15-85b2-fb70f5db6529)

Got the same error of: no route found for POST /admin/albums/new (TeacherWeb.Router) 

I'm on :
elixir 1.17.3-otp-27
    {:phoenix, "~> 1.7.10"},
      {:phoenix_ecto, "~> 4.4"},
      {:ecto_sql, "~> 3.10"},
      {:postgrex, ">= 0.0.0"},
      {:phoenix_html, "~> 4.1.1"},
      {:phoenix_live_reload, "~> 1.2", only: :dev},
      {:phoenix_live_view, "~> 0.20.1"},
      {:floki, ">= 0.30.0", only: :test},
      {:phoenix_live_dashboard, "~> 0.8.2"},
      {:esbuild, "~> 0.8", runtime: Mix.env() == :dev},
      {:tailwind, "~> 0.2.0", runtime: Mix.env() == :dev},
      {:swoosh, "~> 1.3"},
      {:finch, "~> 0.13"},
      {:telemetry_metrics, "~> 0.6"},
      {:telemetry_poller, "~> 1.0"},
      {:gettext, "~> 0.20"},
      {:jason, "~> 1.2"},
      {:dns_cluster, "~> 0.1.1"},
      {:plug_cowboy, "~> 2.5"},
      {:backpex, "~> 0.8.2"},

**Reply by Flo0807** [reply]
Hey @Hkalani! As this seems to be a testing project, can you make it public / push it to GitHub?

**Reply by Hkalani** [reply]
https://github.com/Hkalani/194-backpex

**Comment by Flo0807** [comment]
Hey all!

I figured out why you might get the "no route found for POST XYZ" error when submitting a form.

The reason is that something is wrong with your Live View setup. You may not notice this at first because the server still responds with the HTML content and everything seems fine, even though the Websocket connection to the server is failing.

When the Live View connection fails, the form sends a POST request to the server instead of a Live View / Websocket event. This causes the "no route found" error because there is no such route.

@Hkalani I noticed the following error in the browser console:

```
app.js:26 Uncaught ReferenceError: Alpine is not defined
    at app.js:26:17
    at app.js:56:21
```

You forgot to import Alpine into your app.js file. Since the Live View setup is also configured in the `app.js` file and there is an error, it did not work properly.

Add this line at the top of your `app.js` file and everything should work fine:

```
import Alpine from 'alpinejs
```
